packages/library-utils/src/library_account_type.rs

```diff
@@ -0,0 +1,273 @@
+use std::str::FromStr;
+
+use cosmwasm_schema::cw_serde;
+use cosmwasm_std::{Addr, StdError, StdResult};
+
+use crate::Id;
+
+pub const LIBRARY_ACCOUNT_RAW_PLACEHOLDER: &str = "|lib_acc_placeholder|";
+
+/// An helper type that is used to associate an account or library with an id
+/// When a program is not instantiated yet, ids will be used to reference accounts and libraries
+/// When a program is instantiated, the ids will be replaced by the instantiated addresses
+#[cw_serde]
+#[derive(Eq, PartialOrd, Ord)]
+pub enum LibraryAccountType {
+    #[serde(rename = "|library_account_addr|", alias = "library_account_addr")]
+    Addr(String),
+    #[serde(rename = "|account_id|", alias = "account_id")]
+    AccountId(Id),
+    #[serde(rename = "|library_id|", alias = "library_id")]
+    LibraryId(Id),
+}
+
+impl LibraryAccountType {
+    /// Returns the address as string if it is an address, otherwise returns an error
+    pub fn to_string(&self) -> StdResult<String> {
+        match self {
+            LibraryAccountType::Addr(addr) => Ok(addr.to_string()),
+            LibraryAccountType::AccountId(_) | LibraryAccountType::LibraryId(_) => Err(
+                StdError::generic_err("LibraryAccountType must be an address"),
+            ),
+        }
+    }
+
+    /// Returns the address as Addr if it is an address, otherwise returns an error
+    pub fn to_addr(&self, api: &dyn cosmwasm_std::Api) -> StdResult<Addr> {
+        match self {
+            LibraryAccountType::Addr(addr) => api.addr_validate(addr),
+            LibraryAccountType::AccountId(_) | LibraryAccountType::LibraryId(_) => Err(
+                StdError::generic_err("LibraryAccountType must be an address"),
+            ),
+        }
+    }
+
+    /// There are cases where a library config expects a string, but we still want to use the
+    /// id replacement functionality of the manager.
+    /// Using this function will use a placeholder that can be replaced by the manager
+    /// to the instantiated address
+    pub fn to_raw_placeholder(&self) -> String {
+        let value = match self {
+            // If its an address, we can use it directly
+            LibraryAccountType::Addr(addr) => return addr.to_string(),
+            LibraryAccountType::AccountId(id) => id.to_string(),
+            LibraryAccountType::LibraryId(_) => {
+                panic!("Only accounts can use raw_placeholder functionality")
```

Comment:
maybe worth to return `StdResult` here as well to avoid `panic` &  be consistent with `to_string() / to_addr()`?

Reply:
The reason `to_string() / to_addr()` is using `StdResult` is because both are being used in contract environment.

`to_raw_placeholder()` should never be used in a contract only while building the program, there is no way to recover from this error.

I know it looks weird but as far as I know, this is actually following best practice.